test/sa_stdin.c

```diff
@@ -0,0 +1,67 @@
+#include "../src/sa.h"
+
+char *read_from_stdin() {
+    char buf[1024];
+    size_t string_length = 1024;
+    char *string = malloc(sizeof(char) * 1024);
+    string[0] = '\0';
+    while(fgets(buf, 1024, stdin) && buf[0] != '\n')
+    {
+        string_length += strlen(buf);
+        string = realloc(string, string_length);
+        strcat(string, buf);
```

Comment:
Growing the string this way takes $O(n^2)$. If you always make sure to grow the string by a fixed faction, e.g. length $n$ to $fn$ where $f>0$, then it takes $O(n)$.